mlem/contrib/kubernetes/base.py

```diff
@@ -0,0 +1,300 @@
+import os
+from typing import ClassVar, List, Optional, Tuple
+
+from kubernetes import client, config
+from pydantic import parse_obj_as
+
+from mlem.config import project_config
+from mlem.core.base import MlemABC
+from mlem.core.errors import DeploymentError, EndpointNotFound, MlemError
+from mlem.core.objects import (
+    DeployState,
+    DeployStatus,
+    MlemBuilder,
+    MlemDeployment,
+    MlemEnv,
+    MlemModel,
+)
+from mlem.runtime.client import Client, HTTPClient
+from mlem.runtime.server import Server
+from mlem.ui import EMOJI_OK, echo
+
+from ..docker.base import (
+    DockerDaemon,
+    DockerImage,
+    DockerRegistry,
+    generate_docker_container_name,
+)
+from .build import build_k8s_docker
+from .context import K8sYamlBuildArgs, K8sYamlGenerator
+from .utils import create_k8s_resources, namespace_deleted, pod_is_running
+
+POD_STATE_MAPPING = {
+    "Pending": DeployStatus.STARTING,
+    "Running": DeployStatus.RUNNING,
+    "Succeeded": DeployStatus.STOPPED,
+    "Failed": DeployStatus.CRASHED,
+    "Unknown": DeployStatus.UNKNOWN,
+}
+
+
+def find_index(nodes_list, node_name):
+    for i, each_node in enumerate(nodes_list):
+        if each_node.metadata.name == node_name:
+            return i
+    return -1
+
+
+class ServiceType(MlemABC):
+
+    abs_name: ClassVar = "k8s_service_type"
+    type: ClassVar = "kubernetes"
+
+    class Config:
+        type_root = True
+
+    def get_host_and_port(
+        self, service, namespace="mlem"  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[str], Optional[int]]:
+        """Returns host and port for the service in Kubernetes"""
+        return None, None
+
+
+class NodePortService(ServiceType):
+
+    type: ClassVar = "nodeport"
+
+    def get_host_and_port(self, service, namespace="mlem"):
+        host, port = None, None
+        port = service.items[0].spec.ports[0].node_port
+        node_name = (
+            client.CoreV1Api()
+            .list_namespaced_pod(namespace)
+            .items[0]
+            .spec.node_name
+        )
+        node_list = client.CoreV1Api().list_node().items
+        node_index = find_index(node_list, node_name)
+        if node_index == -1:
+            raise DeploymentError(
+                f"Couldn't find the node where pods in namespace {namespace} exists"
+            )
+        address_dict = node_list[node_index].status.addresses
+        external_ip_found = False
+        for each_address in address_dict:
+            if each_address.type == "ExternalIP":
+                host = each_address.address
```

Comment:
you can just return here and raise instead of current return

Reply:
done